### ISSUE TITLE
Cross-testing fails with rattler-build 0.7.0

### ISSUE THREAD

**Issue by pavelzw** [prefix-dev/rattler-build#487]
This is working with rattler-build `0.6.*` with `--target-platform=linux-64` (when on `osx-arm64`)

```yml
# yaml-language-server: $schema=https://raw.githubusercontent.com/prefix-dev/recipe-format/main/schema.json
context:
  version: '0.0.1'

package:
  name: test-package
  version: ${{ version }}

build:
  number: 0
  script:
    - echo "Hello world!"

test:
  commands:
    - echo "Hello testing world!"
```

This is failing with rattler-build `0.7.0` with `--target-platform=linux-64` (when on `osx-arm64`)

```yml
# yaml-language-server: $schema=https://raw.githubusercontent.com/prefix-dev/recipe-format/main/schema.json
context:
  version: '0.0.1'

package:
  name: test-package
  version: ${{ version }}

build:
  number: 0
  script:
    - echo "Hello world!"

tests:
  - script:
    - echo "Hello testing world!"
```

```
Resolving environment for:

  Platform: linux-64
  Channels: 
   - /home/runner/work/test-project/test-project/output
   - conda-forge
  Specs:
   - test-package ==0.0.1 he8cfe8b_0


2024-01-12T15:02:47.157476Z ERROR rattler_repodata_gateway::fetch: error=repodata not found
Error:   × Failed to setup test environment: Cannot solve the request because of: No
  │ candidates were found for test-package ==0.0.1 he8cfe8b_0.
  │ 
  ╰─▶ Cannot solve the request because of: No candidates were found for test-
      package ==0.0.1 he8cfe8b_0.
```

We should add unit tests to make sure this doesn't break again.

**Comment by 0xbe7a** [comment]
git bisect yielded:

```
➜  rattler-build git:(ebd2afb) git bisect good
b5ac7efdba2748ea8cddd6f79da6eb7f22d4d367 is the first bad commit
commit b5ac7efdba2748ea8cddd6f79da6eb7f22d4d367
Author: Wolf Vollprecht <w.vollprecht@gmail.com>
Date:   Fri Jan 5 10:50:21 2024 +0100

    feat: multiple tests and execution (#418)

 docs/index.md                                      |  16 +-
 docs/recipe_file.md                                | 245 ++++---
 docs/testing.md                                    | 113 ++--
 examples/rich/recipe.yaml                          |  11 +-
 examples/xtensor/recipe.yaml                       |  26 +-
 mkdocs.yml                                         |  12 +-
 rust-tests/src/lib.rs                              |  14 +-
 src/build.rs                                       |  30 +-
 src/lib.rs                                         |   4 +-
 src/main.rs                                        |  34 +-
 src/package_test.rs                                | 733 +++++++++++++++++++++
 src/packaging.rs                                   | 210 +++---
 src/recipe/parser.rs                               |  45 +-
 ...build__recipe__parser__test__test__parsing.snap |   9 +
 src/recipe/parser/test.rs                          | 430 +++++++++---
 ...ler_build__recipe__parser__tests__it_works.snap | 360 ----------
 ...ild__recipe__parser__tests__recipe_windows.snap |  33 +-
 ..._build__recipe__parser__tests__unix_recipe.snap | 367 +++++++++++
 ...er_build__metadata__test__read_full_recipe.snap | 123 ++--
 src/source/copy_dir.rs                             |  32 +-
 src/test.rs                                        | 635 ------------------
 test-data/recipes/flask/recipe.yaml                |  20 +-
 .../package-content-tests/llama-recipe.yaml        |  10 +-
 .../package-content-tests/recipe-test-fail.yaml    |  10 +-
 .../package-content-tests/recipe-test-succeed.yaml |  12 +-
 .../recipes/package-content-tests/rich-recipe.yaml |  20 +-
 .../recipes/test-execution/recipe-test-fail.yaml   |   4 +-
 .../test-execution/recipe-test-succeed.yaml        |   4 +-
 test-data/recipes/test-sources/recipe.yaml         |  27 +-
 test-data/rendered_recipes/rich_recipe.yaml        | 117 ++--
 30 files changed, 2115 insertions(+), 1591 deletions(-)
 create mode 100644 src/package_test.rs
 create mode 100644 src/recipe/parser/snapshots/rattler_build__recipe__parser__test__test__parsing.snap
 delete mode 100644 src/recipe/snapshots/rattler_build__recipe__parser__tests__it_works.snap
 create mode 100644 src/recipe/snapshots/rattler_build__recipe__parser__tests__unix_recipe.snap
 delete mode 100644 src/test.rs
 ```

**Comment by wolfv** [comment]
I think it's just that we set the wrong target_platform during testing?

**Comment by 0xbe7a** [comment]
> I think it's just that we set the wrong target_platform during testing?

No

```
pub async fn run_test(package_file: &Path, config: &TestConfiguration) -> Result<(), TestError> {
    let tmp_repo = tempfile::tempdir()?;
    let target_platform = config.target_platform.unwrap_or_else(Platform::current);

    println!("target platform: {:?}", target_platform);
```

prints

```
Running tests
target platform: Linux64
Removing previously cached package "/Users/belastoyan/Library/Caches/rattler/cache/pkgs/test-package-0.0.1-hb0f4dca_0"
```

I'm not sure what rattler-build is supposed to do here, since nothing provides an installable version (not a foreign architecture, we just built a linux-64 package) of the exact package we just built for osx-arm64. How does conda-build handle this?

**Comment by wolfv** [comment]
@0xbe7a there was a bug though (#491) - I think that restores the previous state of things and is much more correct. 

Unfortunately I don't think I'll be able to add a test today.

I'll try to implement also the "build env" for tests next week and also add a test.

**Comment by pavelzw** [comment]
Closed by #491
